File: scripts/render-pdf-pages.mjs
/**
 * Render vector PDF pages (AutoCAD / Revit drawings) to JPEG.
 *
 * The AutoCAD deliverables carry no embedded images — they are pure vector, so
 * `extract-pdf-images.mjs` finds nothing in them. These need real rasterising.
 *
 * Rasterising is done by `scripts/render_pdf_pages.py` (pypdfium2 + Pillow);
 * this file is the manifest and the runner, so both extraction paths are
 * driven the same way.
 *
 *   node scripts/render-pdf-pages.mjs [--force]
 *
 * Install once:  python -m pip install pypdfium2 pillow
 */

import { spawnSync } from "node:child_process";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const ROOT = join(dirname(fileURLToPath(import.meta.url)), "..");

/**
 * `trim` removes the blank paper around the drawing — these sheets place a
 * small plan on a large page, so an untrimmed render is mostly white.
 */
const MANIFEST = [
  {
    slug: "plans-habitation",
    out: "01.jpg",
    pdf: "AutoCAD/Habitation/Plan D_aménagement villa.pdf",
    page: 0,
    trim: true,
  },
  {
    slug: "plans-habitation",
    out: "02.jpg",
    pdf: "AutoCAD/Habitation/Plan D_aménagement.pdf",
    page: 0,
    trim: true,
  },
  {
    slug: "plans-habitation",
    out: "03.jpg",
    pdf: "AutoCAD/Habitation/Facade (2).pdf",
    page: 0,
    trim: true,
  },
  {
    slug: "maison-de-culture",
    out: "01.jpg",
    pdf: "AutoCAD/Maison de culture/plan d_aménagement.pdf",
    page: 0,
    trim: true,
  },
  {
    slug: "maison-de-culture",
    out: "02.jpg",
    pdf: "AutoCAD/Maison de culture/Facade principale.pdf",
    page: 0,
    trim: true,
  },
  {
    slug: "maison-de-culture",
    out: "03.jpg",
    pdf: "AutoCAD/Maison de culture/COUPE AA.pdf",
    page: 0,
    trim: true,
  },
];

const args = [
  join(ROOT, "scripts", "render_pdf_pages.py"),
  JSON.stringify(MANIFEST),
];
if (process.argv.includes("--force")) args.push("--force");

const res = spawnSync("python", args, { stdio: "inherit", cwd: ROOT });
if (res.status !== 0) {
  console.error(
    "\nRendering failed. Ensure the Python deps are present:\n" +
      "  python -m pip install pypdfium2 pillow",
  );
  process.exit(res.status ?? 1);
}

File: scripts/render_pdf_pages.py
"""Rasterise vector PDF pages for the portfolio.

Driven by scripts/render-pdf-pages.mjs, which passes the manifest as JSON.
Requires: python -m pip install pypdfium2 pillow
"""

import json
import os
import sys

import pypdfium2 as pdfium
from PIL import Image, ImageChops, ImageOps

SRC = os.path.join("public", "DIS STUDIO")
OUT = os.path.join("public", "projets")
MAX_W = 1920


def trim_whitespace(img, pad=28):
    """Crop the blank paper around a drawing.

    These sheets sit a small plan on a large page, so an untrimmed render is
    mostly white. Compare against the corner pixel rather than pure white —
    scans and exports are rarely exactly #FFFFFF.
    """
    rgb = img.convert("RGB")
    bg = Image.new("RGB", rgb.size, rgb.getpixel((0, 0)))
    diff = ImageChops.difference(rgb, bg)
    # Amplify so faint linework still registers against the background.
    bbox = ImageChops.add(diff, diff, 2.0, -24).getbbox()
    if not bbox:
        return img
    left, top, right, bottom = bbox
    left = max(0, left - pad)
    top = max(0, top - pad)
    right = min(img.width, right + pad)
    bottom = min(img.height, bottom + pad)
    if right - left < 80 or bottom - top < 80:
        return img
    return img.crop((left, top, right, bottom))


def main():
    manifest = json.loads(sys.argv[1])
    force = "--force" in sys.argv

    written = skipped = 0
    for item in manifest:
        src = os.path.join(SRC, item["pdf"])
        if not os.path.exists(src):
            print(f"  MISSING  {item['pdf']}")
            continue

        dest_dir = os.path.join(OUT, item["slug"])
        os.makedirs(dest_dir, exist_ok=True)
        dest = os.path.join(dest_dir, item["out"])
        if os.path.exists(dest) and not force:
            skipped += 1
            continue

        pdf = pdfium.PdfDocument(src)
        page = pdf[item.get("page", 0)]
        img = page.render(scale=2.2).to_pil()

        if item.get("trim", True):
            img = trim_whitespace(img)

        # Flatten onto white so any alpha does not turn black in the JPEG.
        if img.mode in ("RGBA", "LA", "P"):
            img = img.convert("RGBA")
            flat = Image.new("RGB", img.size, (255, 255, 255))
            flat.paste(img, mask=img.split()[-1])
            img = flat
        else:
            img = img.convert("RGB")

        if img.width > MAX_W:
            img = ImageOps.contain(img, (MAX_W, 10_000), Image.LANCZOS)

        img.save(dest, quality=84, optimize=True, progressive=True)
        written += 1
        print(
            f"  {item['slug']}/{item['out']}  {img.width}x{img.height}  "
            f"{os.path.getsize(dest) // 1024}KB"
        )

    print(f"\nrendered {written}, skipped {skipped} (use --force to overwrite)")


if __name__ == "__main__":
    main()
